File: solutions/R/regression_foundations_ex3.R

set.seed(42)
n <- 500

age <- round(runif(n, 35, 75))
male <- rbinom(n, 1, 0.5)
smoker <- rbinom(n, 1, 0.25)
cholesterol <- round(rnorm(n, 220, 40))

# Generate CHD outcome (binary) based on logistic model
log_odds <- -7 + 0.06 * age + 0.5 * male + 0.4 * smoker + 0.008 * cholesterol
prob_chd <- plogis(log_odds) # inverse logit
chd <- rbinom(n, 1, prob_chd)

df <- data.frame(
  age,
  male = factor(male),
  smoker = factor(smoker),
  cholesterol,
  chd
)

cat("CHD prevalence:", round(mean(chd), 3), "\n\n")

# Fit logistic regression
model <- glm(
  chd ~ age + male + smoker + cholesterol,
  data = df,
  family = binomial
)
summary(model)

# Odds ratios with 95% CI
or_table <- data.frame(
  OR = exp(coef(model)),
  Lower = exp(confint(model))[, 1],
  Upper = exp(confint(model))[, 2]
)
cat("\nOdds Ratios:\n")
print(round(or_table[-1, ], 3)) # Exclude intercept

# Predicted probability for a specific patient
new_patient <- data.frame(
  age = 60,
  male = factor(1),
  smoker = factor(1),
  cholesterol = 260
)
pred_prob <- predict(model, newdata = new_patient, type = "response")
cat(
  "\nPredicted CHD probability for 60-year-old male smoker with",
  "cholesterol 260:",
  round(pred_prob, 3),
  "\n"
)

File: solutions/python/regression_foundations_ex3.py

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
from scipy.special import expit

np.random.seed(42)
n = 500

age = np.random.uniform(35, 75, n).round()
male = np.random.binomial(1, 0.5, n)
smoker = np.random.binomial(1, 0.25, n)
cholesterol = np.random.normal(220, 40, n).round()

# Generate CHD outcome
log_odds = -7 + 0.06 * age + 0.5 * male + 0.4 * smoker + 0.008 * cholesterol
prob_chd = expit(log_odds)
chd = np.random.binomial(1, prob_chd)

df = pd.DataFrame({
    'age': age, 'male': male, 'smoker': smoker,
    'cholesterol': cholesterol, 'chd': chd
})

print(f"CHD prevalence: {chd.mean():.3f}\n")

# Fit logistic regression
model = smf.logit('chd ~ age + male + smoker + cholesterol', data=df).fit()
print(model.summary())

# Odds ratios with 95% CI
params = model.params[1:]  # exclude intercept
conf = model.conf_int().iloc[1:]
or_table = pd.DataFrame({
    'OR': np.exp(params),
    'Lower 95% CI': np.exp(conf[0]),
    'Upper 95% CI': np.exp(conf[1])
})
print("\nOdds Ratios:")
print(or_table.round(3))

# Predicted probability for a specific patient
new_patient = pd.DataFrame({
    'age': [60], 'male': [1], 'smoker': [1], 'cholesterol': [260]
})
pred_prob = model.predict(new_patient)
print(f"\nPredicted CHD probability for 60yo male smoker, chol=260: {pred_prob.values[0]:.3f}")
